sound/soc/sof/sof-ipc-test-client.c

```diff
@@ -6,22 +6,271 @@
 //
 
 #include <linux/completion.h>
+#include <linux/debugfs.h>
+#include <linux/ktime.h>
 #include <linux/mod_devicetable.h>
 #include <linux/module.h>
 #include <linux/pm_runtime.h>
+#include <linux/slab.h>
+#include <linux/uaccess.h>
 #include <linux/virtual_bus.h>
+#include <sound/sof/header.h>
 #include "sof-client.h"
 
+#define MAX_IPC_FLOOD_DURATION_MS 1000
+#define MAX_IPC_FLOOD_COUNT 10000
+#define IPC_FLOOD_TEST_RESULT_LEN 512
 #define SOF_IPC_CLIENT_SUSPEND_DELAY_MS 3000
 
-/* TODO: dummy probe/remove callbacks for now */
+struct sof_ipc_client_data {
+	struct dentry *dfs_root;
+	char *buf;
+};
+
+static int sof_debug_ipc_flood_test(struct sof_client_dev *cdev,
+				    bool flood_duration_test,
+				    unsigned long ipc_duration_ms,
+				    unsigned long ipc_count)
+{
+	struct sof_ipc_client_data *ipc_client_data = cdev->data;
+	struct device *dev = &cdev->vdev.dev;
+	struct sof_ipc_cmd_hdr hdr;
+	struct sof_ipc_reply reply;
+	u64 min_response_time = U64_MAX;
+	ktime_t start, end, test_end;
+	u64 avg_response_time = 0;
+	u64 max_response_time = 0;
+	u64 ipc_response_time;
+	int i = 0;
+	int ret;
+
+	/* configure test IPC */
+	hdr.cmd = SOF_IPC_GLB_TEST_MSG | SOF_IPC_TEST_IPC_FLOOD;
+	hdr.size = sizeof(hdr);
+
+	/* set test end time for duration flood test */
+	if (flood_duration_test)
+		test_end = ktime_get_ns() + ipc_duration_ms * NSEC_PER_MSEC;
+
+	/* send test IPC's */
+	while (1) {
+		start = ktime_get();
+		ret = sof_client_ipc_tx_message(cdev, hdr.cmd,
+						&hdr, hdr.size, &reply,
+						sizeof(reply));
+		end = ktime_get();
```

Comment:
you could define `start` and `end` inside the loop block, then it would be clear, that they aren't used outside, so you also could measure `end` after checking the return value.

Reply:
@lyakh sorry I didnt understand the ask here. We need the end time irrespective of what the return value is, isnt it?

Reply:
@ranj063 I meant you could move
```
ktime_t start, end;
```
inside the `while` (or `for`) loop. That would make it clear, that they aren't used outside of the loop. And since `end` isn't used outside the loop, if you `break` out of the loop, it isn't needed, right? So,
```
	start = ktime_get();
	ret = sof_client_ipc_tx_message(cdev, hdr.cmd,
					&hdr, hdr.size, &reply,
					sizeof(reply));
	if (ret < 0)
		break;
	end = ktime_get();
```
would work too - you don't need to call `ktime_get()` if you're anyway about to terminate the loop.

Reply:
done